File: server/src/intTest/java/org/tctalent/server/repository/db/PartnerRepositoryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getPartner;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedCountry;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedPartner;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.tctalent.server.model.db.Country;
import org.tctalent.server.model.db.PartnerImpl;
import org.tctalent.server.model.db.Status;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;

public class PartnerRepositoryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private PartnerRepository repo;
  @Autowired
  private CountryRepository countryRepository;
  private PartnerImpl partner;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    partner = getSavedPartner(repo);
  }

  @Test
  public void testGetNamesForIds() {
    PartnerImpl newPartner = getPartner();
    newPartner.setName("TEST PARTNER");
    repo.save(newPartner);
    List<String> names = repo.getNamesForIds(List.of(newPartner.getId(), partner.getId()));
    assertNotNull(names);
    assertFalse(names.isEmpty());
    assertTrue(names.stream().allMatch(name -> name.contains("TEST")));
    // also test ordering as it needs to be asc
    assertEquals(newPartner.getName(), "TEST PARTNER");
  }

  @Test
  public void testFindByStatusOrderByName() {
    partner.setName("ZIGGY");
    partner.setStatus(Status.deleted);
    repo.save(partner);

    PartnerImpl newPartner = getPartner();
    newPartner.setName("AZZY");
    newPartner.setStatus(Status.deleted);
    repo.save(newPartner);

    List<PartnerImpl> results = repo.findByStatusOrderByName(Status.deleted);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(2, results.size());
    assertEquals(newPartner, results.getFirst());
  }

  @Test
  public void testFindByAbbreviation() {
    Optional<PartnerImpl> result = repo.findByAbbreviation(partner.getAbbreviation());
    assertTrue(result.isPresent());
    assertEquals(partner.getId(), result.get().getId());
  }

  @Test
  public void testFindByAbbreviationNone() {
    Optional<PartnerImpl> result = repo.findByAbbreviation("NONE");
    assertFalse(result.isPresent());
  }

  @Test
  public void testFindByDefaultSourcePartner() {
    Optional<PartnerImpl> result = repo.findByDefaultSourcePartner(true);
    assertTrue(result.isPresent());
    assertNotEquals(partner.getId(), result.get().getId());
  }

  @Test
  public void testFindSourcePartnerByAutoassignableCountry() {
    Country country = getSavedCountry(countryRepository);
    partner.setSourceCountries(new HashSet<>(Collections.singletonList(country)));
    repo.save(partner);

    List<PartnerImpl> result = repo.findSourcePartnerByAutoassignableCountry(country);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(1, result.size());
    assertEquals(1, result.getFirst().getSourceCountries().size());
    assertEquals(country.getId(), result.getFirst().getSourceCountries().iterator().next().getId());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/PartnerRepositoryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.jvm.optionals.getOrNull
import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.PartnerImpl
import org.tctalent.server.model.db.Status
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getPartner
import org.tctalent.server.repository.db.integrationhelp.getSavedCountry
import org.tctalent.server.repository.db.integrationhelp.getSavedPartner

class PartnerRepositoryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: PartnerRepository
  @Autowired lateinit var countryRepository: CountryRepository
  private lateinit var partner: PartnerImpl

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    partner = getSavedPartner(repo)
  }

  @Test
  fun `test get names for ids`() {
    val newPartner = getPartner().apply { name = "TEST PARTNER" }
    repo.save(newPartner)
    val names = repo.getNamesForIds(listOf(newPartner.id, partner.id))
    assertNotNull(names)
    assertTrue { names.isNotEmpty() }
    assertTrue { names.all { it.contains("TEST") } }
    // also test ordering as it needs to be asc
    assertEquals(newPartner.name, "TEST PARTNER")
  }

  @Test
  fun `test find by status order by name`() {
    repo.save(
      partner.apply {
        name = "ZIGGY"
        status = Status.deleted
      }
    )
    val newPartner =
      getPartner().apply {
        name = "AZZY"
        status = Status.deleted
      }
    repo.save(newPartner)

    val results = repo.findByStatusOrderByName(Status.deleted)
    assertNotNull(results)
    assertTrue(results.isNotEmpty())
    assertEquals(2, results.count())
    assertEquals(newPartner, results[0])
  }

  @Test
  fun `test find by abbreviation`() {
    val result = repo.findByAbbreviation(partner.abbreviation).getOrNull()
    assertNotNull(result)
    assertEquals(partner.id, result.id)
  }

  @Test
  fun `test find by abbreviation none`() {
    val result = repo.findByAbbreviation("NONE").getOrNull()
    assertNull(result)
  }

  @Test
  fun `test find by default source partner`() {
    val result = repo.findByDefaultSourcePartner(true).getOrNull()
    assertNotNull(result)
    assertNotEquals(partner.id, result.id)
  }

  @Test
  fun `test find source partner by auto assignable country`() {
    val country = getSavedCountry(countryRepository)
    repo.save(partner.apply { sourceCountries = mutableSetOf(country) })

    val result = repo.findSourcePartnerByAutoassignableCountry(country)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    assertEquals(1, result.size)
    assertEquals(1, result.first().sourceCountries.size)
    assertEquals(country.id, result.first().sourceCountries.first().id)
  }
}
